pulsar-common/src/main/java/org/apache/pulsar/common/net/ServiceURI.java

```diff
@@ -0,0 +1,205 @@
+import java.util.stream.Collectors;
+import lombok.AccessLevel;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+import org.apache.commons.lang.StringUtils;
+
+/**
+ * ServiceURI represents service uri within pulsar cluster.
+ *
+ * <p>This file is based on
+ * {@link https://github.com/apache/bookkeeper/blob/master/bookkeeper-common/src/main/java/org/apache/bookkeeper/common/net/ServiceURI.java}
+ */
+@RequiredArgsConstructor(access = AccessLevel.PRIVATE)
+@Getter
+@EqualsAndHashCode
+public class ServiceURI {
+
+    private static final String BINARY_SERVICE = "pulsar";
+    private static final String HTTP_SERVICE = "http";
+    private static final String HTTPS_SERVICE = "https";
+    private static final String SSL_SERVICE = "ssl";
+
+    private static final int BINARY_PORT = 6650;
+    private static final int BINARY_TLS_PORT = 6651;
+    private static final int HTTP_PORT = 8080;
+    private static final int HTTPS_PORT = 8443;
+
+    /**
+     * Create a service uri instance from a uri string.
+     *
+     * @param uriStr service uri string
+     * @return a service uri instance
+     * @throws NullPointerException if {@code uriStr} is null
+     * @throws IllegalArgumentException if the given string violates RFC&nbsp;2396
+     */
+    public static ServiceURI create(String uriStr) {
+        checkNotNull(uriStr, "service uri string is null");
+
+        // a service uri first should be a valid java.net.URI
+        URI uri = URI.create(uriStr);
+
+        return create(uri);
+    }
+
+    /**
+     * Create a service uri instance from a {@link URI} instance.
+     *
+     * @param uri {@link URI} instance
+     * @return a service uri instance
+     * @throws NullPointerException if {@code uriStr} is null
+     * @throws IllegalArgumentException if the given string violates RFC&nbsp;2396
+     */
+    public static ServiceURI create(URI uri) {
+        checkNotNull(uri, "service uri instance is null");
+
+        String serviceName;
+        final String[] serviceInfos;
+        String scheme = uri.getScheme();
+        if (null != scheme) {
+            scheme = scheme.toLowerCase();
+            final String serviceSep = "+";
+            String[] schemeParts = StringUtils.split(scheme, serviceSep);
+            serviceName = schemeParts[0];
+            serviceInfos = new String[schemeParts.length - 1];
+            System.arraycopy(schemeParts, 1, serviceInfos, 0, serviceInfos.length);
+        } else {
+            serviceName = null;
+            serviceInfos = new String[0];
+        }
+
+        String userAndHostInformation = uri.getAuthority();
+        checkArgument(!Strings.isNullOrEmpty(userAndHostInformation),
+            "authority component is missing in service uri : " + uri);
+
+        String serviceUser;
+        List<String> serviceHosts;
+        int atIndex = userAndHostInformation.indexOf('@');
+        Splitter splitter = Splitter.on(CharMatcher.anyOf(",;"));
```

Comment:
comma doesn't work in a URI. so it has to be semicolon. but other systems tends to be using comma. so the idea in ServiceURI is to cover both cases.

Reply:
What you are saying rings a bell, but comma does work in URI. The URI rfc even says it should work (it's a sub-delim)

```
jshell> import java.net.URI

jshell> new URI("http://adfdaf,asfddsaf,adfafd:44")
$2 ==> http://adfdaf,asfddsaf,adfafd:44

jshell> $2.getHost()                                                                                                                                                                                               $3 ==> null
                                                                                                                                                                                                                   jshell> $2.getAuthority()
$4 ==> "adfdaf,asfddsaf,adfafd:44"                                                                                                                                                                                  
```

Reply:
Can’t remember it is URI or URL.

Reply:
https://github.com/apache/bookkeeper/issues/1756#issuecomment-433768776 seems to indicate that comma works for him. And the last comment says semi-colon works, so it seems they both work. His issue was that he was using, full URIs joined by commas. 

I guess it's fine that they're both possible, but we should only document one for avoid confusion. 

I also tried multiple hosts, with ports with URL. this does not work, because of the extra colons. I think that's where we hit this issue in the past, and it was colon we substitutes with semi-colon. This may not have even been bookkeeper. It was many years ago

Reply:
> seems to indicate that comma works for him.

”looks format zk://server1:port,server2:port,server3:port/ledgers does not work“. 

I think you already figured it out. it doesn't work with ports. that's the problem we hit in distributedlog before, so we end up supporting both ',' and ';' in the URI.

Reply:
```
jshell> new URL("http://fooo:22;fo:33/f")
|  Exception java.net.MalformedURLException: Error at index 2 in: "22;fo:33"
|        at URL.<init> (URL.java:661)
|        at URL.<init> (URL.java:523)
|        at URL.<init> (URL.java:470)
|        at (#15:1)
|  Caused by: java.lang.NumberFormatException: Error at index 2 in: "22;fo:33"
|        at NumberFormatException.forCharSequence (NumberFormatException.java:81)
|        at Integer.parseInt (Integer.java:735)
|        at URLStreamHandler.parseURL (URLStreamHandler.java:223)
|        at URL.<init> (URL.java:656)
```
; doesn't work either for separating hosts though, it works for separating host from port

Reply:
try URI.

Reply:
https://github.com/apache/pulsar/blob/272725a9ca4d0334df54aff4b751105949d2c253/pulsar-common/src/test/java/org/apache/pulsar/common/net/ServiceURITest.java#L127

there are test cases covered this. so I would be surprised if URI doesn't work with this format.

Reply:
ya ya, URI works, both with comma and semicolon and each host having a port specified. I was  saying that i think what i remember having the problem was with URL and ports specified.